### PR TITLE
Correct focus in Objects Explorer (vs2017 and up)

### LINKED ISSUE #9311: Object explorer not working correctly in Visual Studio 2017 (v15.3 and up)
### Steps to reproduce:

1. Open any type of project with object explorer available, such _.NET_.
1. Open object explorer with <kbd>ALT+CTRL+j</kbd>.
1. Ensure you're in the objects tree and select the first expandable item.
1. Expand it by pressing <kbd>right arrow</kbd>.
1. Enter into it with <kbd>right</kbd> or <kbd>down</kbd> arrow.
1. Contract it by pressing <kbd>left arrow</kbd> twice.
1. Press the <kbd>END</kbd> key to select the last item on the tree.
1. Repeat the last four steps with this last one.

### Actual behavior:

When entering into the expanded item, the next item gains the focus. It occurs only if the expanded item is not the last.

If the expanded item is the last, it works as expected.

### Expected behavior:

Expected behavior is like any other tree view even under visual studio, like the solution explorer.

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

alpha-16813,a0b991af

#### Windows version:

Version	10.0.17763 build 17763

#### Name and version of other software in use when reproducing the issue:

Visual Studio Community 15.9.7

#### Does the issue still occur after restarting your PC?

Yes, reboot doesn't change anything.

#### Have you tried any other versions of NVDA?

I've used all stable versions of nvda between 2017.3 and this alpha I'm using now.

### PR DESCRIPTION
**Pull request by francipvb**
### Link to issue number:
Closes #9311 
### Summary of the issue:
In visual studio 2017 (version 15.3 and up), the object explorer window is not working properly due to an incorrect MSAA implementation.
See #9313 for a previous approach to solve this.
### Description of how this pull request fixes the issue:
This adds an overlay class for the tree view items in object explorer. It takes a `focusRedirect` event and checks for its states
If `STATE_FOCUSED` is not in the states set, it redirects the event to the real focused object.
I'm not really sure if it is the right way of doing this.
### Testing performed:
Tested against visual studio 15.9 and up to 16.0.rc4.
### Known issues with pull request:
State changes are not reported for secondary items, so expand and contract events are not reported by NVDA. I don't have more information.
### Change log entry:

Section: Bug fixes

* In visual studio 2017 and up, in *Objects Explorer* window, selected item in objects tree or members tree with categories is reported correctly again. (#9311)

### CODE REVIEW

**Comment by francipvb** [comment]
cc @leonardder

**Comment by codeofdusk** [comment]
Could you maybe use the focus redirection system instead of getting `event_gainFocus`?

**Comment by lukaszgo1** [comment]
@francipvb You're importing ui in the appModule hovewer it isn't used anywhere - probably left when debugging.

**Comment by francipvb** [comment]
Hello,

> 
> 
> Could you maybe use the focus redirection system instead of getting `event_gainFocus`?

Thanks, done. It is a more clean approach 👍 

> 
> 
> @francipvb You're importing ui in the appModule hovewer it isn't used anywhere - probably left when debugging.

Fixed too, thanks. 👍 

Cheers,

**Comment by leonardder** [comment]
@dantegagne: Do you think there could be a possibility to fix this in Visual Studio itself?

File: source/appModules/devenv.py [hunk 1]
```diff
@@ -88,6 +90,9 @@
 
 class AppModule(appModuleHandler.AppModule):
 	def chooseNVDAObjectOverlayClasses(self, obj, clsList):
+		vs_major, vs_minor, rest = self.productVersion.split(".", 2)
```
Comment (leonardder) [inline source/appModules/devenv.py]:
```suggestion
		vsMajor, vsMinor = map(int, self.productVersion.split(".", 2)[:2])
```

You can also remove the lina below. Note the change from vs_major to vsMajor, vs_minor to vsMinor, which is more like the general code style of NVDA.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Just some small things

**Review by LeonarddeR — APPROVED** [review]
I think this is ok from a code perspective. I need to have a look with Visual Studio, probably next week.
In the meanwhile, could you please update the initial description of the pr to reflect the current implementation (i.e. with focusRedirect instead of patching event_gainFocus)?

**Review by feerrenrut — APPROVED** [review]
@leonardder I'm going to merge this in, would you mind reporting back here if there are any issues with the focus in the object explorer window.

**Pull request merged**